File: src/Tests/StateUITests/WireFuzzTests.swift
// Trying to BREAK the decoders rather than to read them.
//
// The four channels the HOST writes and this side reads - an act's reply, an
// event's payload, an event the host raised by name, an environment push -
// and the hydration a kept state starts from. Every buffer, every truncation,
// every single-byte change, and every buffer through every decoder, because a
// decoder handed a shape that is not its own must answer nil exactly as it
// answers a mangled one.
//
// The statement is sharper here than on the C# side, because the failure is:
// an index past the end of a Swift array is a TRAP, which takes the process
// down rather than raising something a caller could answer. So what is claimed
// is that a decoder always RETURNS - and the suite dying where it stands IS
// the failure. The attempt COUNTS are asserted so that a loop which stopped
// early cannot pass as a loop which found nothing.

import Foundation
import XCTest
@testable import StateUI

final class WireFuzzTests: XCTestCase {
    /// A buffer to break, and the decoder whose shape it really is.
    private struct Sample {
        let name: String
        let bytes: [UInt8]
        let read: @Sendable ([UInt8]) -> Bool
    }

    /// Every decoder, by the name a failure would have to be traced back to.
    /// Each answers whether it read the buffer, which is what makes the wrong
    /// shape and the mangled buffer one statement.
    private static let decoders: [(name: String, read: @Sendable ([UInt8]) -> Bool)] = [
        ("decodePayload", { Wire.decodePayload($0) != nil }),
        ("decodeReply", { Wire.decodeReply($0) != nil }),
        ("decodeHostEvent", { Wire.decodeHostEvent($0) != nil }),
        ("decodeEnvironment", { Wire.decodeEnvironment($0) != nil }),
        ("decodePersistent", { Wire.decodePersistent($0) != nil }),
    ]

    /// What a single byte is changed by: every bit, the low bit alone, and the
    /// high bit alone - enough to reach a length's top byte, a tag and a count.
    private static let changes: [UInt8] = [0xFF, 0x01, 0x80]

    /// The payload fixtures, each paired with its own decoder, plus a
    /// hydration buffer - which has no fixture of its own because no message
    /// carries it, and would otherwise be the one channel nothing here breaks.
    private func corpus() throws -> [Sample] {
        let root = Fixtures.directory.appendingPathComponent("payloads")
        var samples: [Sample] = []

        for name in try FileManager.default
            .contentsOfDirectory(atPath: root.path)
            .filter({ $0.hasSuffix(".bin") })
            .sorted()
        {
            let bytes = [UInt8](try Data(contentsOf: root.appendingPathComponent(name)))
            let read: @Sendable ([UInt8]) -> Bool

            if name.hasPrefix("host-event") {
                read = { Wire.decodeHostEvent($0) != nil }
            } else if name.hasPrefix("environment") {
                read = { Wire.decodeEnvironment($0) != nil }
            } else if name.hasPrefix("reply") {
                read = { Wire.decodeReply($0) != nil }
            } else {
                read = { Wire.decodePayload($0) != nil }
            }

            samples.append(Sample(name: name, bytes: bytes, read: read))
        }

        var hydration: [UInt8] = []
        hydration.u8(Wire.version)
        hydration.u16(2)
        hydration.string("test.count")
        hydration.value(.number(4))
        hydration.string("test.loud")
        hydration.value(.bool(true))

        samples.append(Sample(
            name: "hydration", bytes: hydration,
            read: { Wire.decodePersistent($0) != nil }))

        // A corpus that quietly emptied would leave every test here passing
        // while breaking nothing.
        XCTAssertGreaterThan(samples.count, 10)

        return samples
    }

    /// Every buffer the host writes, read whole by its own decoder - the floor
    /// the two tests below stand on, since a corpus of buffers that no decoder
    /// accepts would be broken in every direction already.
    func testEveryPayloadIsReadWholeByItsOwnDecoder() throws {
        for sample in try corpus() {
            XCTAssertTrue(sample.read(sample.bytes), "\(sample.name) did not read whole")
        }
    }

    /// A decoder handed a buffer of somebody else's shape answers rather than
    /// trapping. Nothing sends one - the host writes each channel into its own
    /// entry point - so this is about what happens when something does.
    func testEveryDecoderAnswersForAShapeThatIsNotItsOwn() throws {
        var attempts = 0

        for sample in try corpus() {
            for decoder in Self.decoders {
                _ = decoder.read(sample.bytes)
                attempts += 1
            }
        }

        XCTAssertEqual(attempts, try corpus().count * Self.decoders.count)
    }

    /// A buffer that stops early is refused, wherever it stops. Every proper
    /// prefix of every buffer: each decoder bounds-checks its way forward AND
    /// requires the last value to end at the last byte, so no prefix reads as
    /// a smaller, plausible payload.
    ///
    /// An EMPTY buffer is the exception and not a truncation at all: it is
    /// what the host crosses for an event with nothing to say, which is why
    /// `decodePayload` answers it with an empty list.
    func testEveryProperPrefixOfEveryPayloadIsRefused() throws {
        for sample in try corpus() {
            for length in 1..<sample.bytes.count {
                XCTAssertFalse(
                    sample.read(Array(sample.bytes.prefix(length))),
                    "\(sample.name) cut to \(length) of \(sample.bytes.count) bytes was read")
            }
        }
    }

    /// A length no buffer could hold is refused rather than believed. It
    /// crosses UNSIGNED and four bytes wide, and is compared that way: `Int`
    /// is 32 bits on a 32-bit target, so the widest lengths do not fit the one
    /// a length is finally used as. Nothing writes such a length; this is what
    /// happens when something does.
    func testALengthNoBufferCouldHoldIsRefused() {
        var buffer: [UInt8] = []
        buffer.u8(Wire.version)
        buffer.u16(1)
        buffer.append(contentsOf: [0xFF, 0xFF, 0xFF, 0xFF])

        XCTAssertNil(Wire.decodePersistent(buffer))
    }

    /// Every single-byte change in every buffer is answered - read or refused,
    /// and the tree that comes out of a changed value is nobody's business.
    /// What is claimed is only that the decoder returns at all.
    func testEverySingleByteChangeIsAnsweredRatherThanTrapped() throws {
        let samples = try corpus()
        var attempts = 0
        var expected = 0

        for sample in samples {
            expected += sample.bytes.count * Self.changes.count

            for at in sample.bytes.indices {
                for change in Self.changes {
                    var changed = sample.bytes
                    changed[at] ^= change
                    _ = sample.read(changed)
                    attempts += 1
                }
            }
        }

        XCTAssertEqual(attempts, expected)
    }
}

File: _old/src/Tests/StateUIRuntime.Tests/WireFuzzTests.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// Trying to BREAK the reader rather than to read it.
//
// Every fixture the Swift side wrote, put through every truncation and every
// single-byte change there is. The invariant is NOT that a mangled message is
// refused - a changed byte inside a number is still a number, and the tree
// that comes out is nobody's business - but that the only way this reader ever
// fails is InvalidDataException. That is the one failure the two callers are
// written to survive: `StateUISession.Render` says so to the reader, and
// `PerformCommands` cashes the receipt so every act in an unreadable batch
// fails back to the handler awaiting it. Anything else escaping a reader is a
// failure mode nothing was designed against.
//
// The other direction - the replies and payloads THIS side writes and Swift
// reads - is fuzzed in Swift, in WireFuzzTests.swift.

using StateUI.Runtime.Protocol;

namespace StateUI.Runtime.Tests;

public class WireFuzzTests
{
    /// <summary>
    /// What a single byte is changed by: every bit, the low bit alone, and the
    /// high bit alone. Three is enough to reach the interesting places - a
    /// length's top byte, a tag, a count - without turning a one-second suite
    /// into a minute of it.
    /// </summary>
    private static readonly byte[] Changes = [0xFF, 0x01, 0x80];

    /// <summary>
    /// The render fixtures: every `.bin` except the command batches, which are
    /// read by the other entry point, and the payloads, which this side writes.
    /// </summary>
    private static List<string> Messages() => Corpus(under: null);

    /// <summary>The command batches, read by <see cref="SwiftWire.ReadCommands"/>.</summary>
    private static List<string> Commands() => Corpus(under: "commands");

    private static List<string> Corpus(string? under)
    {
        var found = Directory
            .EnumerateFiles(Fixtures.Directory, "*.bin", SearchOption.AllDirectories)
            .Select(path => Path.GetRelativePath(Fixtures.Directory, path).Replace('\\', '/'))
            .Where(name => under is null
                ? !name.StartsWith("commands/") && !name.StartsWith("payloads/")
                : name.StartsWith(under + "/"))
            .OrderBy(name => name, StringComparer.Ordinal)
            .ToList();

        // A corpus that quietly emptied - a moved directory, a renamed suffix -
        // would leave every test here passing while reading nothing.
        Assert.NotEmpty(found);

        return found;
    }

    /// <summary>
    /// Reads, and says what happened: null when the bytes were read, the
    /// exception when they were not. An InvalidDataException is the contract
    /// and is simply absorbed; anything else is named with the byte that
    /// caused it, because that is the only way back to it.
    /// </summary>
    private static void Refused(Action read, bool mustRefuse, string what)
    {
        bool wasRead = false;

        try
        {
            read();
            wasRead = true;
        }
        catch (InvalidDataException)
        {
        }
        catch (Exception ex)
        {
            Assert.Fail($"{what} failed with {ex.GetType().Name}: {ex.Message}");
        }

        if (mustRefuse && wasRead)
        {
            Assert.Fail($"{what} was read as if it were whole");
        }
    }

    /// <summary>
    /// A message that stops early is refused, wherever it stops. Every proper
    /// prefix of every fixture, which is what makes this a statement about the
    /// format rather than about one hand-cut example: the reader bounds-checks
    /// each step AND requires the root to end exactly at the last byte, so
    /// there is no prefix that reads as a smaller, plausible tree.
    /// </summary>
    [Fact]
    public void EveryProperPrefixOfEveryMessageIsRefused()
    {
        foreach (string name in Messages())
        {
            byte[] whole = Fixtures.ReadBytes(name);

            for (int length = 0; length < whole.Length; length++)
            {
                byte[] cut = whole.AsSpan(0, length).ToArray();

                Refused(
                    () => SwiftWire.ReadMessage(cut, new SwiftWireDictionary()),
                    mustRefuse: true,
                    $"{name} cut to {length} of {whole.Length} bytes");
            }
        }
    }

    /// <summary>
    /// Every single-byte change in every fixture is either read or refused,
    /// and never anything else. This is the one that reaches the arithmetic:
    /// a string's length is four bytes wide, so the top byte alone decides
    /// whether the reader is asked for a length no buffer could hold.
    /// </summary>
    [Fact]
    public void EverySingleByteChangeInEveryMessageIsReadOrRefused()
    {
        foreach (string name in Messages())
        {
            byte[] whole = Fixtures.ReadBytes(name);

            for (int at = 0; at < whole.Length; at++)
            {
                foreach (byte change in Changes)
                {
                    byte[] changed = (byte[])whole.Clone();
                    changed[at] ^= change;

                    Refused(
                        () => SwiftWire.ReadMessage(changed, new SwiftWireDictionary()),
                        mustRefuse: false,
                        $"{name} with byte {at} of {whole.Length} changed by 0x{change:X2}");
                }
            }
        }
    }

    /// <summary>
    /// A length no buffer could hold is refused rather than believed. It
    /// crosses UNSIGNED and four bytes wide, so the widest of them does not
    /// fit the int a length is finally used as - and the bounds check has to
    /// happen before that narrowing, or the reader goes on to ask for a
    /// negative slice and fails with something no caller is written for.
    /// </summary>
    [Fact]
    public void ALengthNoBufferCouldHoldIsRefused()
    {
        byte[] bytes =
        [
            SwiftWire.Version,
            1,                          // complete
            0, 0, 0, 0,                 // generation 0
            1, 0,                       // one announcement
            1, 0,                       // id #1
            0xFF, 0xFF, 0xFF, 0xFF,     // whose name is four gigabytes long
        ];

        Assert.Throws<InvalidDataException>(
            () => SwiftWire.ReadMessage(bytes, new SwiftWireDictionary()));
    }

    /// <summary>
    /// A tree that nests deeper than any real one is refused as bad bytes,
    /// rather than taking the process down with it.
    /// </summary>
    /// <remarks>
    /// The reader walks children by recursion, so a length field naming
    /// thousands of levels would exhaust the stack - which is not an exception
    /// anything can catch, and so not a failure the session can report. A real
    /// page is tens of levels deep.
    /// </remarks>
    [Fact]
    public void ATreeNestedDeeperThanAnyRealOneIsRefused()
    {
        const string type = "VerticalStackLayout";

        var bytes = new List<byte>
        {
            SwiftWire.Version,
            1,                          // complete
            0, 0, 0, 0,                 // generation 0
            1, 0,                       // one announcement
            1, 0,                       // under id #1
        };

        bytes.AddRange([(byte)type.Length, 0, 0, 0]);
        bytes.AddRange(System.Text.Encoding.UTF8.GetBytes(type));

        // Each level is a whole, well-formed node: a numbered identity, the
        // type announced above, and one arranged child. Only the DEPTH is
        // wrong, so nothing else can be what refuses it.
        for (int level = 0; level < 2000; level++)
        {
            bytes.AddRange([1, (byte)(level & 0x7F), 0, 0, 0]);
            bytes.AddRange([1, 0]);
            bytes.AddRange([5, 1, 0]);
        }

        bytes.AddRange([1, 0, 0, 0, 0]);
        bytes.AddRange([1, 0]);
        bytes.Add(0);

        InvalidDataException refused = Assert.Throws<InvalidDataException>(
            () => SwiftWire.ReadMessage([.. bytes], new SwiftWireDictionary()));

        Assert.Contains("nests deeper", refused.Message);
    }

    /// <summary>
    /// A list too long for the field that counts it is refused where it is
    /// WRITTEN, naming the list and the limit.
    /// </summary>
    /// <remarks>
    /// The counts are one byte on the host's own channels, and a registered
    /// act or event is free to answer with as many values as it likes. A plain
    /// cast would truncate the count, the far side would then refuse the
    /// buffer over its leftover bytes, and the failure would read as a native
    /// library and a runtime built from different versions.
    /// </remarks>
    [Fact]
    public void APayloadWithMoreValuesThanTheCountCanSayIsRefused()
    {
        SwiftWireValue[] many =
            [.. Enumerable.Range(0, 256).Select(number => SwiftWireValue.Of((double)number))];

        ArgumentException refused = Assert.Throws<ArgumentException>(
            () => SwiftWire.WritePayload(many));

        Assert.Contains("256", refused.Message);
        Assert.Contains("values in one payload", refused.Message);

        // And the one below the limit is written, so the refusal is about the
        // limit rather than about a list this side cannot write at all.
        Assert.NotNull(SwiftWire.WritePayload([.. many.Take(255)]));
    }

    /// <summary>
    /// The same two statements about a command batch, whose reader is the one
    /// an awaiting handler is hanging off.
    /// </summary>
    [Fact]
    public void EveryProperPrefixOfEveryCommandBatchIsRefused()
    {
        foreach (string name in Commands())
        {
            byte[] whole = Fixtures.ReadBytes(name);

            for (int length = 0; length < whole.Length; length++)
            {
                byte[] cut = whole.AsSpan(0, length).ToArray();

                Refused(
                    () => SwiftWire.ReadCommands(cut, new SwiftWireDictionary()),
                    mustRefuse: true,
                    $"{name} cut to {length} of {whole.Length} bytes");
            }
        }
    }

    [Fact]
    public void EverySingleByteChangeInEveryCommandBatchIsReadOrRefused()
    {
        foreach (string name in Commands())
        {
            byte[] whole = Fixtures.ReadBytes(name);

            for (int at = 0; at < whole.Length; at++)
            {
                foreach (byte change in Changes)
                {
                    byte[] changed = (byte[])whole.Clone();
                    changed[at] ^= change;

                    Refused(
                        () => SwiftWire.ReadCommands(changed, new SwiftWireDictionary()),
                        mustRefuse: false,
                        $"{name} with byte {at} of {whole.Length} changed by 0x{change:X2}");
                }
            }
        }
    }
}
